consensus/tendermint/tmdb.go

```diff
@@ -0,0 +1,283 @@
+// It panics if the DB scan fails.
+func (s *tendermintDB[V, H, A]) getWALCount(height height) walIter {
+	prefix := tmdb.WALEntry.Key(encodeHeight(height))
+	count := walIter(0)
+
+	err := s.db.View(func(snap db.Snapshot) error {
+		iter, err := snap.NewIterator(prefix, false)
+		if err != nil {
+			return err
+		}
+		defer iter.Close()
+
+		if !iter.Seek(prefix) {
+			return nil // No entries for this height
+		}
+		for ; iter.Valid(); iter.Next() {
+			key := iter.Key()
+			if !bytes.HasPrefix(key, prefix) {
+				break
+			}
+			count++
+		}
+		return nil
+	})
+	if err != nil {
+		// Failing to retrieve the WAL msgs can result in a loss of funds (consensus slashing)
+		panic(fmt.Sprintf("getWALCount: failed to scan WAL entries for height %d: %v", height, err))
+	}
+	return count
+}
+
+// DeleteWALMsgs iterates through the expected message keys based on the stored count.
+// Note: This operates on the batch. Changes are only persisted after CommitBatch() is called.
+func (s *tendermintDB[V, H, A]) DeleteWALMsgs(height height) error {
+	numMsgs, ok := s.walCount[height]
+	if !ok {
+		return nil
+	}
+
+	heightBytes := encodeHeight(height)
+
+	startIterBytes := encodeNumMsgsAtHeight(walIter(1))
+	endIterBytes := encodeNumMsgsAtHeight(walIter(uint32(numMsgs + 1)))
+
+	startKey := tmdb.WALEntry.Key(heightBytes, startIterBytes)
+	endKey := tmdb.WALEntry.Key(heightBytes, endIterBytes)
+	if err := s.batch.DeleteRange(startKey, endKey); err != nil {
+		return fmt.Errorf("DeleteWALMsgs: failed to add delete range [%x, %x) to batch: %w", startKey, endKey, err)
+	}
+
+	delete(s.walCount, height)
+	return nil
+}
+
+// SetWALEntry implements TMDBInterface.
+func (s *tendermintDB[V, H, A]) SetWALEntry(entry IsWALMsg, height height) error {
+	wrapper := walEntry[V, H, A]{
+		Type:  entry.msgType(),
+		Entry: entry,
+	}
+	wrappedEntry, err := cbor.Marshal(wrapper)
+	if err != nil {
+		return fmt.Errorf("SetWALEntry: marshal wrapper failed: %w", err)
+	}
+
+	numMsgsAtHeight, ok := s.walCount[height]
+	if !ok {
+		numMsgsAtHeight = 0
+	}
+	nextNumMsgsAtHeight := numMsgsAtHeight + 1
+	nextNumMsgsAtHeightBytes := encodeNumMsgsAtHeight(nextNumMsgsAtHeight)
+
+	msgKey := tmdb.WALEntry.Key(encodeHeight(height), nextNumMsgsAtHeightBytes)
+	if err := s.batch.Put(msgKey, wrappedEntry); err != nil {
+		return fmt.Errorf("writeWALEntryToBatch: failed to set MsgsAtHeight: %w", err)
+	}
+
+	s.walCount[height] = nextNumMsgsAtHeight
+	return nil
+}
+
+// GetWALMsgs implements TMDBInterface.
+func (s *tendermintDB[V, H, A]) GetWALMsgs(height height) ([]walEntry[V, H, A], error) {
+	rawEntries, err := s.scanWALRaw(height)
+	if err != nil {
+		return nil, fmt.Errorf("GetWALMsgs: failed to scan raw WAL entries for height %d: %w", height, err)
+	}
+
+	walMsgs := make([]walEntry[V, H, A], len(rawEntries))
+
+	for i, rawEntry := range rawEntries {
+		var walMsg walEntry[V, H, A]
+		if err := cbor.Unmarshal(rawEntry, &walMsg); err != nil {
```

Comment:
I think this is slightly more efficient
```suggestion
		if err := cbor.Unmarshal(rawEntry, &walMsgs[i]); err != nil {
```